src/gil.rs

```diff
@@ -168,74 +168,56 @@ impl Drop for GILGuard {
 
 /// Thread-safe storage for objects which were inc_ref / dec_ref while the GIL was not held.
 struct ReferencePool {
-    locked: AtomicBool,
-    pointers_to_incref: UnsafeCell<Vec<NonNull<ffi::PyObject>>>,
-    pointers_to_decref: UnsafeCell<Vec<NonNull<ffi::PyObject>>>,
-}
-
-struct Lock<'a> {
-    lock: &'a AtomicBool,
-}
-
-impl<'a> Lock<'a> {
-    fn new(lock: &'a AtomicBool) -> Self {
-        while lock.compare_and_swap(false, true, Ordering::Acquire) {
-            spin_loop_hint();
-        }
-        Self { lock }
-    }
-}
-
-impl<'a> Drop for Lock<'a> {
-    fn drop(&mut self) {
-        self.lock.store(false, Ordering::Release);
-    }
+    pointers_to_incref: Mutex<Vec<NonNull<ffi::PyObject>>>,
+    pointers_to_decref: Mutex<Vec<NonNull<ffi::PyObject>>>,
 }
 
 impl ReferencePool {
     const fn new() -> Self {
         Self {
-            locked: AtomicBool::new(false),
-            pointers_to_incref: UnsafeCell::new(Vec::new()),
-            pointers_to_decref: UnsafeCell::new(Vec::new()),
+            pointers_to_incref: const_mutex(Vec::new()),
+            pointers_to_decref: const_mutex(Vec::new()),
         }
     }
 
     fn register_incref(&self, obj: NonNull<ffi::PyObject>) {
-        let _lock = Lock::new(&self.locked);
-        let v = self.pointers_to_incref.get();
-        unsafe { (*v).push(obj) };
+        self.pointers_to_incref.lock().push(obj)
     }
 
     fn register_decref(&self, obj: NonNull<ffi::PyObject>) {
-        let _lock = Lock::new(&self.locked);
-        let v = self.pointers_to_decref.get();
-        unsafe { (*v).push(obj) };
+        self.pointers_to_decref.lock().push(obj)
     }
 
     fn update_counts(&self, _py: Python) {
-        let _lock = Lock::new(&self.locked);
+        macro_rules! swap_vec_with_lock {
+            // Get vec from one of ReferencePool's mutexes via lock, swap vec if needed, unlock.
+            ($cell:expr) => {{
+                let mut locked = $cell.lock();
+                let mut out = None;
```

Comment:
I don't think we need Option here, too. Zero-sized Vec is internally a constant and fast.

Reply:
Yep agreed :+1: if you want to change this before I get a chance later tonight, please push to this branch :smile:

Reply:
Done :+1: